File: src/harness_core/dispatch/evaluation_stub.py
"""Evaluation stub: basic boundary checks, no quality judgment."""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

from .dispatch_decision import DispatchDecision
from .executor_adapter import ExecutionResult

# ---------------------------------------------------------------------------
# Schema version
# ---------------------------------------------------------------------------

EVALUATION_RESULT_SCHEMA_VERSION = "evaluation_result.v1"

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

EVAL_CHECK_NAMES: tuple[str, ...] = (
    "schema_validity",
    "boundary_compliance",
    "output_present",
    "error_free",
    "human_review_required",
)


# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class EvaluationCheck:
    check_id: str
    name: str
    status: str  # "pass" | "fail" | "warning" | "skipped"
    reason: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "check_id": self.check_id,
            "name": self.name,
            "status": self.status,
            "reason": self.reason,
        }


@dataclass(frozen=True)
class EvaluationResult:
    evaluation_id: str
    dispatch_id: str
    decision_id: str
    execution_result_id: str
    status: str  # "pass" | "fail" | "needs_human_review" | "not_evaluated"
    checks: tuple[EvaluationCheck, ...]
    created_at: str
    quality_score: float | None = None
    requires_retry: bool = False
    retry_reason: str | None = None
    schema_version: str = EVALUATION_RESULT_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "evaluation_id": self.evaluation_id,
            "dispatch_id": self.dispatch_id,
            "decision_id": self.decision_id,
            "execution_result_id": self.execution_result_id,
            "status": self.status,
            "checks": [c.to_dict() for c in self.checks],
            "quality_score": self.quality_score,
            "requires_retry": self.requires_retry,
            "retry_reason": self.retry_reason,
            "created_at": self.created_at,
        }


# ---------------------------------------------------------------------------
# Stub evaluator
# ---------------------------------------------------------------------------


class EvaluationStub:
    """Basic boundary checks. No quality judgment, no LLM judge."""

    def evaluate(
        self, result: ExecutionResult, decision: DispatchDecision
    ) -> EvaluationResult:
        checks = [
            self._check_schema_validity(result),
            self._check_boundary_compliance(result, decision),
            self._check_output_present(result),
            self._check_error_free(result),
            self._check_human_review_required(result, decision),
        ]

        failed = [c for c in checks if c.status == "fail"]
        warnings = [c for c in checks if c.status == "warning"]

        if failed:
            status = "fail"
        elif any(c.name == "human_review_required" and c.status == "warning" for c in checks):
            status = "needs_human_review"
        elif warnings:
            status = "pass"
        else:
            status = "pass"

        return EvaluationResult(
            evaluation_id=f"eval-{uuid.uuid4().hex[:12]}",
            dispatch_id=result.dispatch_id,
            decision_id=result.decision_id,
            execution_result_id=result.result_id,
            status=status,
            checks=tuple(checks),
            created_at=datetime.now(timezone.utc).isoformat(),
        )

    def _check_schema_validity(self, result: ExecutionResult) -> EvaluationCheck:
        valid = bool(result.result_id and result.dispatch_id and result.decision_id)
        return EvaluationCheck(
            check_id=f"chk-{uuid.uuid4().hex[:8]}",
            name="schema_validity",
            status="pass" if valid else "fail",
            reason="required fields present" if valid else "missing required fields",
        )

    def _check_boundary_compliance(
        self, result: ExecutionResult, decision: DispatchDecision
    ) -> EvaluationCheck:
        if result.executor_type == "provider":
            return EvaluationCheck(
                check_id=f"chk-{uuid.uuid4().hex[:8]}",
                name="boundary_compliance",
                status="fail",
                reason="provider executor not allowed in Phase 1",
            )
        return EvaluationCheck(
            check_id=f"chk-{uuid.uuid4().hex[:8]}",
            name="boundary_compliance",
            status="pass",
            reason=f"executor_type={result.executor_type} within Phase 1 boundaries",
        )

    def _check_output_present(self, result: ExecutionResult) -> EvaluationCheck:
        if result.executor_type == "noop":
            return EvaluationCheck(
                check_id=f"chk-{uuid.uuid4().hex[:8]}",
                name="output_present",
                status="warning",
                reason="noop executor produces no output (expected)",
            )
        if result.output or result.prompt_pack:
            return EvaluationCheck(
                check_id=f"chk-{uuid.uuid4().hex[:8]}",
                name="output_present",
                status="pass",
                reason="output or prompt_pack present",
            )
        return EvaluationCheck(
            check_id=f"chk-{uuid.uuid4().hex[:8]}",
            name="output_present",
            status="fail",
            reason="no output and no prompt_pack",
        )

    def _check_error_free(self, result: ExecutionResult) -> EvaluationCheck:
        if result.error_domain:
            return EvaluationCheck(
                check_id=f"chk-{uuid.uuid4().hex[:8]}",
                name="error_free",
                status="fail",
                reason=f"error: {result.error_domain}: {result.error_message}",
            )
        return EvaluationCheck(
            check_id=f"chk-{uuid.uuid4().hex[:8]}",
            name="error_free",
            status="pass",
            reason="no errors",
        )

    def _check_human_review_required(
        self, result: ExecutionResult, decision: DispatchDecision
    ) -> EvaluationCheck:
        if result.status == "manual_pending":
            return EvaluationCheck(
                check_id=f"chk-{uuid.uuid4().hex[:8]}",
                name="human_review_required",
                status="warning",
                reason="manual executor awaiting human pasteback",
            )
        if decision.execution_policy.get("requires_human_review"):
            return EvaluationCheck(
                check_id=f"chk-{uuid.uuid4().hex[:8]}",
                name="human_review_required",
                status="warning",
                reason="execution policy requires human review",
            )
        return EvaluationCheck(
            check_id=f"chk-{uuid.uuid4().hex[:8]}",
            name="human_review_required",
            status="pass",
            reason="no human review required",
        )

File: engine/src/evaluation_stub.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::dispatch_decision::DispatchDecision;
use crate::executor_adapter::ExecutionResult;
use crate::runtime::FixtureRuntime;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct EvaluationCheck {
    pub check_id: String,
    pub name: String,
    pub status: String,
    pub reason: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct EvaluationResult {
    pub schema_version: String,
    pub evaluation_id: String,
    pub dispatch_id: String,
    pub decision_id: String,
    pub execution_result_id: String,
    pub status: String,
    pub checks: Vec<EvaluationCheck>,
    pub quality_score: Option<f64>,
    pub requires_retry: bool,
    pub retry_reason: Option<String>,
    pub created_at: String,
}

impl EvaluationResult {
    pub fn to_value(&self) -> Value {
        serde_json::to_value(self).expect("EvaluationResult should serialize")
    }
}

pub trait Evaluator: Send + Sync {
    fn evaluate(
        &self,
        result: &ExecutionResult,
        decision: &DispatchDecision,
        runtime: &mut FixtureRuntime,
    ) -> EvaluationResult;
}

#[derive(Default)]
pub struct EvaluationStub;

impl EvaluationStub {
    pub fn evaluate(
        &self,
        result: &ExecutionResult,
        decision: &DispatchDecision,
        runtime: &mut FixtureRuntime,
    ) -> EvaluationResult {
        let requires_review = decision
            .execution_policy
            .get("requires_human_review")
            .and_then(Value::as_bool)
            .unwrap_or(false);
        let mut checks = vec![
            EvaluationCheck {
                check_id: runtime.id("chk-"),
                name: "schema_validity".to_string(),
                status: "pass".to_string(),
                reason: "required fields present".to_string(),
            },
            EvaluationCheck {
                check_id: runtime.id("chk-"),
                name: "boundary_compliance".to_string(),
                status: "pass".to_string(),
                reason: "executor_type=noop within boundaries".to_string(),
            },
            EvaluationCheck {
                check_id: runtime.id("chk-"),
                name: "output_present".to_string(),
                status: "warning".to_string(),
                reason: "noop executor produces no output (expected)".to_string(),
            },
            EvaluationCheck {
                check_id: runtime.id("chk-"),
                name: "error_free".to_string(),
                status: "pass".to_string(),
                reason: "no errors".to_string(),
            },
        ];
        checks.push(EvaluationCheck {
            check_id: runtime.id("chk-"),
            name: "human_review_required".to_string(),
            status: if requires_review { "warning" } else { "pass" }.to_string(),
            reason: if requires_review {
                "execution policy requires human review"
            } else {
                "no human review required"
            }
            .to_string(),
        });
        if let Some(ref output) = result.output {
            checks.push(EvaluationCheck {
                check_id: runtime.id("chk-"),
                name: "output_non_empty".to_string(),
                status: if output.is_empty() { "fail" } else { "pass" }.to_string(),
                reason: if output.is_empty() {
                    "LLM returned empty output"
                } else {
                    "output is non-empty"
                }
                .to_string(),
            });
            checks.push(EvaluationCheck {
                check_id: runtime.id("chk-"),
                name: "output_substantial".to_string(),
                status: if !output.is_empty() && output.len() < 10 {
                    "warning"
                } else {
                    "pass"
                }
                .to_string(),
                reason: if !output.is_empty() && output.len() < 10 {
                    "Output suspiciously short"
                } else {
                    "output length is acceptable"
                }
                .to_string(),
            });
        }
        if result.status == "failed" {
            checks.push(EvaluationCheck {
                check_id: runtime.id("chk-"),
                name: "no_error_in_output".to_string(),
                status: "fail".to_string(),
                reason: result
                    .error_message
                    .clone()
                    .unwrap_or_else(|| "execution failed".to_string()),
            });
        }
        let has_fail = checks.iter().any(|c| c.status == "fail");
        let status = if requires_review {
            "needs_human_review"
        } else if has_fail {
            "fail"
        } else {
            "pass"
        };
        let should_retry = has_fail && !requires_review;
        EvaluationResult {
            schema_version: "evaluation_result.v1".to_string(),
            evaluation_id: runtime.id("eval-"),
            dispatch_id: result.dispatch_id.clone(),
            decision_id: result.decision_id.clone(),
            execution_result_id: result.result_id.clone(),
            status: status.to_string(),
            checks,
            quality_score: None,
            requires_retry: should_retry,
            retry_reason: if should_retry {
                Some("evaluation check failed".to_string())
            } else {
                None
            },
            created_at: runtime.now(),
        }
    }
}

impl Evaluator for EvaluationStub {
    fn evaluate(
        &self,
        result: &ExecutionResult,
        decision: &DispatchDecision,
        runtime: &mut FixtureRuntime,
    ) -> EvaluationResult {
        EvaluationStub::evaluate(self, result, decision, runtime)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn needs_human_review_prevents_retry() {
        let stub = EvaluationStub;
        let mut runtime = FixtureRuntime::new();
        let result = ExecutionResult {
            schema_version: "execution_result.v1".to_string(),
            result_id: runtime.id("exec-"),
            dispatch_id: "d1".to_string(),
            decision_id: "dec1".to_string(),
            executor_type: "noop".to_string(),
            status: "completed".to_string(),
            output: Some("ok".to_string()),
            prompt_pack: None,
            input_tokens: None,
            output_tokens: None,
            estimated_cost: None,
            latency_ms: None,
            error_domain: None,
            error_message: None,
            provider_request_id: None,
            attempt_number: None,
            finish_reason: None,
            usage_source: None,
            created_at: runtime.now(),
        };
        let decision = DispatchDecision {
            decision_id: "dec1".to_string(),
            execution_policy: serde_json::json!({"requires_human_review": true}),
            ..DispatchDecision::default()
        };
        let eval = stub.evaluate(&result, &decision, &mut runtime);
        assert_eq!(eval.status, "needs_human_review");
        assert!(
            !eval.requires_retry,
            "should not retry when human review is required"
        );
    }
}
